### PR TITLE
rsetboot for redfish support

### PR DESCRIPTION
**Pull request by xuweibj**
### The PR is for task _https://github.com/xcat2/xcat2-task-management/issues/518

_##Feature description##_
rsetboot cd/hd/def/floppy/net/setup/stat [-p]

### The content of the PR:
* [x] _##The basic code logic_
* stat: Get members of Systems and get BootSourceOverrideTarget & BootSourceOverrideEnabled
* hd/net/cd/floppy/setup: Check the value want to set whether in  BootSourceOverrideTarget@Redfish.AllowableValues, if in, send patch request to set. 
* -p: Set BootSourceOverrideEnabled as ``Continuous``
* If set to def, set BootSourceOverrideEnabled as ``Disabled``

### The UT result
```
# rsetboot boston32-redfish hd -p
Thu Dec 20 22:02:55 2018 boston32-redfish: [redfish_debug] login curl -k -X POST -H "Content-Type: application/json" https://50.6.32.1/redfish/v1/SessionService/Sessions -d '{"UserName": "ADMIN", "Password": "xxxxxx"}' -v
Thu Dec 20 22:02:55 2018 boston32-redfish: [redfish_debug] login 201 User Session
Thu Dec 20 22:02:55 2018 boston32-redfish: [redfish_debug] get_members curl -k -X GET -H "Content-Type: application/json" -H "X-Auth-Token: xxxxxx" https://50.6.32.1/redfish/v1/Systems
Thu Dec 20 22:02:55 2018 boston32-redfish: [redfish_debug] get_members 200 Computer System Collection
Thu Dec 20 22:02:55 2018 boston32-redfish: [redfish_debug] get_boot_actions curl -k -X GET -H "Content-Type: application/json" -H "X-Auth-Token: xxxxxx" https://50.6.32.1/redfish/v1/Systems/Computer
Thu Dec 20 22:02:56 2018 boston32-redfish: [redfish_debug] get_boot_actions 200 OpenPOWER Computer System
Thu Dec 20 22:02:56 2018 boston32-redfish: [redfish_debug] set_boot_state curl -k -X PATCH -H "Content-Type: application/json" -H "X-Auth-Token: xxxxxx" https://50.6.32.1/redfish/v1/Systems/Computer {"Boot": {"BootSourceOverrideTarget": "Hdd", "BootSourceOverrideEnabled": "Continuous"}} -v
Thu Dec 20 22:02:56 2018 boston32-redfish: [redfish_debug] set_boot_state 200 Successfully Completed Request
Thu Dec 20 22:02:56 2018 boston32-redfish: [redfish_debug] get_members curl -k -X GET -H "Content-Type: application/json" -H "X-Auth-Token: xxxxxx" https://50.6.32.1/redfish/v1/Systems
Thu Dec 20 22:02:56 2018 boston32-redfish: [redfish_debug] get_members 200 Computer System Collection
Thu Dec 20 22:02:56 2018 boston32-redfish: [redfish_debug] get_boot_state curl -k -X GET -H "Content-Type: application/json" -H "X-Auth-Token: xxxxxx" https://50.6.32.1/redfish/v1/Systems/Computer
Thu Dec 20 22:02:56 2018 boston32-redfish: [redfish_debug] get_boot_state 200 OpenPOWER Computer System
boston32-redfish: Hard Drive
```

```
# rsetboot boston32-redfish def
Thu Dec 20 22:04:56 2018 boston32-redfish: [redfish_debug] login curl -k -X POST -H "Content-Type: application/json" https://50.6.32.1/redfish/v1/SessionService/Sessions -d '{"UserName": "ADMIN", "Password": "xxxxxx"}' -v
Thu Dec 20 22:04:56 2018 boston32-redfish: [redfish_debug] login 201 User Session
Thu Dec 20 22:04:56 2018 boston32-redfish: [redfish_debug] get_members curl -k -X GET -H "Content-Type: application/json" -H "X-Auth-Token: xxxxxx" https://50.6.32.1/redfish/v1/Systems
Thu Dec 20 22:04:57 2018 boston32-redfish: [redfish_debug] get_members 200 Computer System Collection
Thu Dec 20 22:04:57 2018 boston32-redfish: [redfish_debug] get_boot_actions curl -k -X GET -H "Content-Type: application/json" -H "X-Auth-Token: xxxxxx" https://50.6.32.1/redfish/v1/Systems/Computer
Thu Dec 20 22:04:57 2018 boston32-redfish: [redfish_debug] get_boot_actions 200 OpenPOWER Computer System
Thu Dec 20 22:04:57 2018 boston32-redfish: [redfish_debug] set_boot_state curl -k -X PATCH -H "Content-Type: application/json" -H "X-Auth-Token: xxxxxx" https://50.6.32.1/redfish/v1/Systems/Computer {"Boot": {"BootSourceOverrideTarget": "None", "BootSourceOverrideEnabled": "Disabled"}} -v
Thu Dec 20 22:04:57 2018 boston32-redfish: [redfish_debug] set_boot_state 200 Successfully Completed Request
Thu Dec 20 22:04:57 2018 boston32-redfish: [redfish_debug] get_members curl -k -X GET -H "Content-Type: application/json" -H "X-Auth-Token: xxxxxx" https://50.6.32.1/redfish/v1/Systems
Thu Dec 20 22:04:57 2018 boston32-redfish: [redfish_debug] get_members 200 Computer System Collection
Thu Dec 20 22:04:57 2018 boston32-redfish: [redfish_debug] get_boot_state curl -k -X GET -H "Content-Type: application/json" -H "X-Auth-Token: xxxxxx" https://50.6.32.1/redfish/v1/Systems/Computer
Thu Dec 20 22:04:58 2018 boston32-redfish: [redfish_debug] get_boot_state 200 OpenPOWER Computer System
boston32-redfish: boot override inactive
```

### CODE REVIEW

File: xCAT-openbmc-py/lib/python/agent/hwctl/redfish_client.py [hunk 1]
```diff
@@ -119,11 +148,136 @@ def handle_response (self, resp, cmd=''):
...
+        except KeyError as e:
+            raise SelfServerException('Get KeyError %s' % e.message)
+
+    def _get_bmc_actions(self):
+
+        members = self._get_members(MANAGER_URL)
+        target_url = members[0]['@odata.id']
+        data = self.request('GET', target_url, cmd='get_bmc_actions')
+        try:
+            actions = data['Actions']['#Manager.Reset']['ResetType@Redfish.AllowableValues']
```
Comment (gurevichmark) [inline xCAT-openbmc-py/lib/python/agent/hwctl/redfish_client.py]: Would it be better to put strings like "#Manager.Reset" and "ResetType@Redfish.AllowableValues" into variables ?

File: xCAT-openbmc-py/lib/python/agent/hwctl/redfish_client.py [hunk 2]
```diff
@@ -119,11 +148,136 @@ def handle_response (self, resp, cmd=''):
...
+        except KeyError as e:
+            raise SelfServerException('Get KeyError %s' % e.message)
+
+        return (target_url, actions)
+
+    def reboot_bmc(self, optype='warm'):
+
+        target_url, actions = self._get_bmc_actions()
+        if BMC_RESET_TYPE not in actions:
+            raise SelfClientException('Unsupport option: %s' % BMC_RESET_TYPE)
```
Comment (gurevichmark) [inline xCAT-openbmc-py/lib/python/agent/hwctl/redfish_client.py]: Better to say "Unsupported option"

**Comment by zet809** [comment]
Good to me, I agree to merge. Thx @xuweibj

**Comment by whowutwut** [comment]
@xuweibj Why are we doing force push to the branch after reviews started?

**Comment by xuweibj** [comment]
@whowutwut I just modified to add one line code, so just use ``--amend``, will use ``-m`` next time.

**Pull request merged**